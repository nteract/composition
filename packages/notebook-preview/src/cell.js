/* @flow */
/* eslint-disable react/prefer-stateless-function */
import React from "react";
import { List as ImmutableList, Map as ImmutableMap } from "immutable";

import CodeCell from "./code-cell";
import MarkdownCell from "./markdown-cell";

export type CellProps = {
  cell: any,
  displayOrder: ImmutableList<any>,
  id: string,
  language: string,
  theme: string,
  transforms: ImmutableMap<string, any>
};

export class Cell extends React.PureComponent {
  props: CellProps;

  render(): ?React.Element<any> {
    const cell = this.props.cell;
    const type = cell.get("cell_type");

    console.log(this.props.displayOrder);

    return (
      <div className={`cell ${type === "markdown" ? "text" : "code"}`}>
        {type === "markdown"
          ? <MarkdownCell
              cell={cell}
              id={this.props.id}
              theme={this.props.theme}
            />
          : <CodeCell
              cell={cell}
              id={this.props.id}
              theme={this.props.theme}
              language={this.props.language}
              displayOrder={this.props.displayOrder}
              transforms={this.props.transforms}
            />}
      </div>
    );
  }
}

export default Cell;
