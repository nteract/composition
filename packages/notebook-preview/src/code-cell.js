// @flow
import React from "react";
import { List as ImmutableList, Map as ImmutableMap } from "immutable";
import { Display } from "@nteract/display-area";

import Inputs from "./inputs";

import Editor from "./editor";

import LatexRenderer from "./latex";

type Props = {
  cell: ImmutableMap<string, any>,
  displayOrder: ImmutableList<any>,
  id: string,
  language: string,
  theme: string,
  tip: boolean,
  transforms: ImmutableMap<string, any>,
  running: boolean,
  models: ImmutableMap<string, any>
};

class CodeCell extends React.PureComponent {
  props: Props;

  static defaultProps = {
    running: false,
    tabSize: 4,
    models: new ImmutableMap()
  };

  isOutputHidden(): any {
    return this.props.cell.getIn(["metadata", "outputHidden"]);
  }

  isInputHidden(): any {
    return this.props.cell.getIn(["metadata", "inputHidden"]);
  }

  isOutputExpanded() {
    return this.props.cell.getIn(["metadata", "outputExpanded"], true);
  }

  render(): ?React.Element<any> {
    return (
      <div className={this.props && this.props.running ? "cell-running" : ""}>
        {!this.isInputHidden()
          ? <div className="input-container">
              <Inputs
                executionCount={this.props.cell.get("execution_count")}
                running={this.props.running}
              />
              <Editor
                completion
                id={this.props.id}
                input={this.props.cell.get("source")}
                language={this.props.language}
                theme={this.props.theme}
                tip={this.props.tip}
                cellFocused={false}
                onChange={() => {}}
                onFocusChange={() => {}}
                channels={{}}
                cursorBlinkRate={0}
                executionState={"not connected"}
                editorFocused={false}
                focusAbove={() => {}}
                focusBelow={() => {}}
              />
            </div>
          : <div className="input-container invisible" />}
        <LatexRenderer>
          <div className="outputs">
            <Display
              className="outputs-display"
              outputs={this.props.cell.get("outputs")}
              displayOrder={this.props.displayOrder}
              transforms={this.props.transforms}
              theme={this.props.theme}
              tip={this.props.tip}
              expanded={this.isOutputExpanded()}
              isHidden={this.isOutputHidden()}
              models={this.props.models}
            />
          </div>
        </LatexRenderer>
      </div>
    );
  }
}

export default CodeCell;
