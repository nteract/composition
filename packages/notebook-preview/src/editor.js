// @flow
import React from "react";
import CodeMirrorWrapper from "@nteract/editor/lib/wrapper";

type Props = {
  children?: React$Element<any>
};

const EditorView = (props: Props): React$Element<any> =>
  <div className="input">
    {props.children}
  </div>;

export default CodeMirrorWrapper(EditorView, { readOnly: true });
