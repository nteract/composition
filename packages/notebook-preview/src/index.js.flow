// @flow
/* eslint-disable */

declare var NotebookPreview: any;
export default NotebookPreview;