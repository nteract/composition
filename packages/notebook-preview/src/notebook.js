/* @flow */
import React from 'react';

import { List as ImmutableList, Map as ImmutableMap } from 'immutable';

import { displayOrder, transforms } from '@nteract/transforms';

import Cell from './cell';

type Props = {
  displayOrder: ImmutableList<any>,
  notebook: any,
  transforms: ImmutableMap<string, any>,
  theme: string
};

export function getLanguageMode(notebook: any): string {
  // The syntax highlighting language should be set in the language info
  // object.  First try codemirror_mode, then name, and fallback on 'null'.
  const language = notebook.getIn(
    ['metadata', 'language_info', 'codemirror_mode', 'name'],
    notebook.getIn(
      ['metadata', 'language_info', 'codemirror_mode'],
      notebook.getIn(['metadata', 'language_info', 'name'], 'text')
    )
  );
  return language;
}

export class Notebook extends React.PureComponent {
  props: Props;
  createCellElement: (s: string) => ?React.Element<any>;

  static defaultProps = {
    displayOrder,
    transforms
  };

  static contextTypes = {
    store: React.PropTypes.object
  };

  constructor(): void {
    super();
    this.createCellElement = this.createCellElement.bind(this);
  }

  createCellElement(id: string): ?React.Element<any> {
    const cellMap = this.props.notebook.get('cellMap');
    const cell = cellMap.get(id);

    return (
      <div className="cell-container" key={`cell-container-${id}`}>
        <Cell
          key={id}
          id={id}
          cell={cell}
          displayOrder={this.props.displayOrder}
          transforms={this.props.transforms}
          theme={this.props.theme}
          language={getLanguageMode(this.props.notebook)}
        />
      </div>
    );
  }

  render(): ?React.Element<any> {
    if (!this.props.notebook) {
      return <div className="notebook" />;
    }

    const cellOrder = this.props.notebook.get('cellOrder');
    return (
      <div>
        <div className="notebook">
          {cellOrder.map(this.createCellElement)}
        </div>
      </div>
    );
  }
}

export default Notebook;
