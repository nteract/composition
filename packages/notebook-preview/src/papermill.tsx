// @flow
import * as React from "react";

type PapermillMetadata = {
  status?: "pending" | "running" | "completed"
  // TODO: Acknowledge / use other papermill metadata
};

export const PapermillView = (props: PapermillMetadata) => {
  if (!props.status) {
    return null;
  }

  if (props.status === "running") {
    return (
      <React.Fragment>
        <div className="papermill">Executing with Papermill...</div>
        <style jsx>{`
          .papermill {
            width: 100%;
            background-color: #e8f2ff;
            padding-left: 10px;
            padding-top: 1em;
            padding-bottom: 1em;
            padding-right: 0;
            margin-right: 0;
            box-sizing: border-box;
          }
        `}</style>
      </React.Fragment>
    );
  }
  return null;
};
