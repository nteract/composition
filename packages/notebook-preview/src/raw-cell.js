// @flow

import React from "react";

export default class MarkdownCell extends React.PureComponent {
  render(): ?React.Element<any> {
    return (
      <pre>
        {this.props.cell.get("source")}
      </pre>
    );
  }
}
