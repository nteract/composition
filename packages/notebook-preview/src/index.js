/* @flow */
import React from "react";
import { List as ImmutableList, Map as ImmutableMap } from "immutable";

import CommonMark from "commonmark";
import MarkdownRenderer from "commonmark-react-renderer";

import { Display } from "@nteract/display-area";
import { displayOrder, transforms } from "@nteract/transforms";

import { emptyNotebook, appendCellToNotebook } from "@nteract/commutable";
import { createCodeCell } from "@nteract/commutable/lib/structures";

import { _nextgen } from "@nteract/core/components";

import LatexRenderer from "./latex";
import { PapermillView } from "./papermill";

const {
  Cell,
  Input,
  Prompt,
  PromptBuffer,
  Editor,
  Outputs,
  Notebook
} = _nextgen;

const themes = require("@nteract/core/themes");

type MDRender = (input: string) => string;

const parser = new CommonMark.Parser();
const renderer = new MarkdownRenderer();

const mdRender: MDRender = input => renderer.render(parser.parse(input));

type Props = {
  displayOrder: Array<string>,
  notebook: any,
  transforms: Object,
  theme: string
};

export class NotebookPreview extends React.PureComponent<Props> {
  static defaultProps = {
    displayOrder,
    transforms,
    notebook: appendCellToNotebook(
      emptyNotebook,
      createCodeCell().set("source", "test")
    ),
    theme: "light"
  };

  render(): ?React$Element<any> {
    // TODO: Rely on setState to convert notebook from plain JS to commutable format

    const notebook = this.props.notebook;

    // Propagated from the hide_(all)_input nbextension
    const allSourceHidden = notebook.getIn(["metadata", "hide_input"], false);

    const language = notebook.getIn(
      ["metadata", "language_info", "codemirror_mode", "name"],
      notebook.getIn(
        ["metadata", "language_info", "codemirror_mode"],
        notebook.getIn(["metadata", "language_info", "name"], "text")
      )
    );

    const cellOrder = notebook.get("cellOrder");
    const cellMap = notebook.get("cellMap");

    return (
      <div className="notebook-preview">
        <Notebook>
          {cellOrder.map(cellID => {
            const cell = cellMap.get(cellID);
            const cellType = cell.get("cell_type");
            const source = cell.get("source");

            switch (cellType) {
              case "code":
                const sourceHidden =
                  allSourceHidden ||
                  cell.getIn(["metadata", "inputHidden"]) ||
                  cell.getIn(["metadata", "hide_input"]);

                const outputHidden =
                  cell.get("outputs").size === 0 ||
                  cell.getIn(["metadata", "outputHidden"]);

                return (
                  <Cell key={cellID}>
                    <PapermillView
                      {...cell
                        .getIn(["metadata", "papermill"], ImmutableMap())
                        .toJS()}
                    />
                    <Input hidden={sourceHidden}>
                      <Prompt />
                      <Editor language={language} theme={this.props.theme}>
                        {source}
                      </Editor>
                    </Input>
                    <Outputs
                      hidden={outputHidden}
                      expanded={cell.getIn(
                        ["metadata", "outputExpanded"],
                        true
                      )}
                    >
                      <Display
                        outputs={cell.get("outputs").toJS()}
                        transforms={transforms}
                        displayOrder={displayOrder}
                      />
                    </Outputs>
                  </Cell>
                );
              case "markdown":
                return (
                  <Cell key={cellID}>
                    <div className="content-margin">
                      {/* TODO: embed mathjax on the page here, otherwise LaTeX
                              rendering doesn't actually happen
                    */}
                      <LatexRenderer>{mdRender(source)}</LatexRenderer>
                    </div>
                    <style jsx>{`
                      .content-margin {
                        padding-left: calc(var(--prompt-width, 50px) + 10px);
                        padding-top: 10px;
                        padding-bottom: 10px;
                        padding-right: 10px;
                      }
                    `}</style>
                  </Cell>
                );
              case "raw":
                return (
                  <Cell key={cellID}>
                    <pre className="raw-cell">
                      {source}
                      <style jsx>{`
                        raw-cell {
                          background: repeating-linear-gradient(
                            -45deg,
                            transparent,
                            transparent 10px,
                            #efefef 10px,
                            #f1f1f1 20px
                          );
                        }
                      `}</style>
                    </pre>
                  </Cell>
                );

              default:
                return (
                  <Cell key={cellID}>
                    <Outputs>
                      <pre>{`Cell Type "${cellType}" is not implemented`}</pre>
                    </Outputs>
                  </Cell>
                );
            }
          })}
        </Notebook>
        <style>{`:root {
          ${themes[this.props.theme]}
            --cell-shadow-hover-1: none;
            --cell-shadow-hover-2: none;
            --cell-shadow-focus-1: none;
            --cell-shadow-focus-2: none;
            --cell-bg-hover: var(--prompt-bg);
            --cell-bg-focus: var(--prompt-bg);
          }
        `}</style>
      </div>
    );
  }
}

export default NotebookPreview;
