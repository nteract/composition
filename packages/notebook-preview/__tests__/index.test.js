import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

import { displayOrder, transforms } from "@nteract/transforms";

import NotebookPreview from "./../src";

import { dummyCommutable, dummyJSON } from "../../core/src/dummy";

// In order to get reproducable snapshots we need to mock the uuid package
jest.mock("uuid/v4", () => {
  let uuid = 1;
  return jest.fn(() => uuid++);
});

describe("Notebook", () => {
  it("accepts an Immutable.List of cells", () => {
    const component = shallow(
      <NotebookPreview
        notebook={dummyCommutable}
        theme="light"
        tip
        displayOrder={displayOrder}
        transforms={transforms}
      />
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  it("accepts an Object of cells", () => {
    const component = shallow(
      <NotebookPreview
        notebook={dummyJSON}
        theme="light"
        tip
        displayOrder={displayOrder}
        transforms={transforms}
      />
    );
    expect(toJson(component)).toMatchSnapshot();
  });
});
