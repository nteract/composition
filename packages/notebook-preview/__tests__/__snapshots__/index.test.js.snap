// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notebook accepts an Immutable.List of cells 1`] = `
<Provider
  delay={0}
  didFinishTypeset={null}
  input="tex"
  loading={null}
  noGate={false}
  onError={[Function]}
  onLoad={null}
  options={Object {}}
  src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML"
>
  <div
    className="notebook-preview"
  >
    <Cells>
      <Cell
        _hovered={false}
        isSelected={false}
        key="2"
      >
        <div
          className="jsx-2513275823 content-margin"
        >
          <MarkdownRender
            source="## The Notable Nteract Notebook

**It's a notebook!**
"
          />
        </div>
        <JSXStyle
          css=".content-margin.jsx-2513275823{padding-left:calc(var(--prompt-width,50px) + 10px);padding-top:10px;padding-bottom:10px;padding-right:10px;}"
          styleId="2513275823"
        />
      </Cell>
      <Cell
        _hovered={false}
        isSelected={false}
        key="3"
      >
        <PapermillView
          status={null}
        />
        <Input
          hidden={false}
        >
          <Prompt
            blank={false}
            counter={11}
            queued={false}
            running={false}
          />
          <Source
            className="input"
            language="ipython"
            theme="light"
          >
            import IPython

from IPython.display import HTML
from IPython.display import Markdown
from IPython.display import display

display(HTML('&lt;h1&gt;Multiple&lt;/h1&gt;'))
display(HTML('&lt;p&gt;Display Elements&lt;/p&gt;'))
display(Markdown('**awesome**'))

print('hey')
42
          </Source>
        </Input>
        <Outputs
          expanded={true}
          hidden={false}
        >
          <Display
            displayOrder={
              Array [
                "application/vdom.v1+json",
                "application/json",
                "application/javascript",
                "text/html",
                "text/markdown",
                "text/latex",
                "image/svg+xml",
                "image/gif",
                "image/png",
                "image/jpeg",
                "text/plain",
              ]
            }
            expanded={false}
            isHidden={false}
            models={Immutable.Map {}}
            outputs={
              Array [
                Object {
                  "data": Object {
                    "text/plain": "<IPython.core.display.HTML object>",
                  },
                  "metadata": Object {},
                  "output_type": "display_data",
                },
              ]
            }
            theme="light"
            transforms={
              Object {
                "application/javascript": [Function],
                "application/json": [Function],
                "application/vdom.v1+json": [Function],
                "image/gif": [Function],
                "image/jpeg": [Function],
                "image/png": [Function],
                "image/svg+xml": [Function],
                "text/html": [Function],
                "text/latex": [Function],
                "text/markdown": [Function],
                "text/plain": [Function],
              }
            }
          />
        </Outputs>
      </Cell>
    </Cells>
    <style>
      :root {
          

  --theme-app-bg: white;
  --theme-app-fg: var(--nt-color-midnight);
  --theme-app-border: var(--nt-color-grey-light);

  --theme-primary-bg: var(--nt-color-grey-lightest);
  --theme-primary-bg-hover: var(--nt-color-grey-lighter);
  --theme-primary-bg-focus: var(--nt-color-grey-light);

  --theme-primary-fg: var(--nt-color-midnight-light);
  --theme-primary-fg-hover: var(--nt-color-midnight);
  --theme-primary-fg-focus: var(--theme-app-fg);

  --theme-secondary-bg: var(--theme-primary-bg);
  --theme-secondary-bg-hover: var(--theme-primary-bg-hover);
  --theme-secondary-bg-focus: var(--theme-primary-bg-focus);

  --theme-secondary-fg: var(--nt-color-midnight-lighter);
  --theme-secondary-fg-hover: var(--nt-color-midnight-light);
  --theme-secondary-fg-focus: var(--theme-primary-fg);

  --theme-primary-shadow-hover: 1px  1px 3px rgba(0, 0, 0, 0.12), -1px -1px 3px rgba(0, 0, 0, 0.12);
  --theme-primary-shadow-focus: 3px  3px 9px rgba(0, 0, 0, 0.12), -3px -3px 9px rgba(0, 0, 0, 0.12);

  --theme-title-bar-bg: var(--theme-primary-bg-hover);

  --theme-menu-bg: var(--theme-primary-bg);
  --theme-menu-bg-hover: var(--theme-primary-bg-hover);
  --theme-menu-bg-focus: var(--theme-primary-bg-focus);
  --theme-menu-shadow: var(--theme-primary-shadow-hover);

  --theme-menu-fg: var(--theme-app-fg);
  --theme-menu-fg-hover: var(--theme-app-fg);
  --theme-menu-fg-focus: var(--theme-app-fg);

  --theme-cell-bg: var(--theme-app-bg);
  --theme-cell-shadow-hover: var(--theme-primary-shadow-hover);
  --theme-cell-shadow-focus: var(--theme-primary-shadow-focus);

  --theme-cell-prompt-bg: var(--theme-primary-bg);
  --theme-cell-prompt-bg-hover: var(--theme-primary-bg-hover);
  --theme-cell-prompt-bg-focus: var(--theme-primary-bg-focus);

  --theme-cell-prompt-fg: var(--theme-secondary-fg);
  --theme-cell-prompt-fg-hover: var(--theme-secondary-fg-hover);
  --theme-cell-prompt-fg-focus: var(--theme-secondary-fg-focus);

  --theme-cell-toolbar-bg: var(--theme-primary-bg);
  --theme-cell-toolbar-bg-hover: var(--theme-primary-bg-hover);
  --theme-cell-toolbar-bg-focus: var(--theme-primary-bg-focus);

  --theme-cell-toolbar-fg: var(--theme-secondary-fg);
  --theme-cell-toolbar-fg-hover: var(--theme-secondary-fg-hover);
  --theme-cell-toolbar-fg-focus: var(--theme-secondary-fg-focus);

  --theme-cell-menu-bg: var(--theme-primary-bg);
  --theme-cell-menu-bg-hover: var(--theme-primary-bg-hover);
  --theme-cell-menu-bg-focus: var(--theme-primary-bg-focus);

  --theme-cell-menu-fg: var(--theme-primary-fg);
  --theme-cell-menu-fg-hover: var(--theme-primary-fg-hover);
  --theme-cell-menu-fg-focus: var(--theme-primary-fg-focus);

  --theme-cell-input-bg: var(--theme-secondary-bg);
  --theme-cell-input-fg: var(--theme-app-fg);

  --theme-cell-output-bg: var(--theme-app-bg);
  --theme-cell-output-fg: var(--theme-primary-fg);

  --theme-cell-creator-bg: var(--theme-app-bg);

  --theme-cell-creator-fg: var(--theme-secondary-fg);
  --theme-cell-creator-fg-hover: var(--theme-secondary-fg-hover);
  --theme-cell-creator-fg-focus: var(--theme-secondary-fg-focus);

  --theme-pager-bg: #fafafa;

  --cm-background: #fafafa;
  --cm-color: black;

  --cm-gutter-bg: white;

  --cm-comment: #a86;
  --cm-keyword: blue;
  --cm-string: #a22;
  --cm-builtin: #077;
  --cm-special: #0aa;
  --cm-variable: black;
  --cm-number: #3a3;
  --cm-meta: #555;
  --cm-link: #3a3;
  --cm-operator: black;
  --cm-def: black;

  --cm-activeline-bg: #e8f2ff;
  --cm-matchingbracket-outline: grey;
  --cm-matchingbracket-color: black;

  --cm-hint-color: var(--cm-color);
  --cm-hint-color-active: var(--cm-color);
  --cm-hint-bg: var(--theme-app-bg);
  --cm-hint-bg-active: #abd1ff;

  --status-bar: #eeedee;

            --theme-cell-shadow-hover: none;
            --theme-cell-shadow-focus: none;
            --theme-cell-prompt-bg-hover: var(--theme-cell-prompt-bg);
            --theme-cell-prompt-bg-focus: var(--theme-cell-prompt-bg);
            --theme-cell-prompt-fg-hover: var(--theme-cell-prompt-fg);
            --theme-cell-prompt-fg-focus: var(--theme-cell-prompt-fg);
          }
        
    </style>
  </div>
</Provider>
`;

exports[`Notebook accepts an Object of cells 1`] = `
<Provider
  delay={0}
  didFinishTypeset={null}
  input="tex"
  loading={null}
  noGate={false}
  onError={[Function]}
  onLoad={null}
  options={Object {}}
  src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML"
>
  <div
    className="notebook-preview"
  >
    <Cells>
      <Cell
        _hovered={false}
        isSelected={false}
        key="14"
      >
        <div
          className="jsx-2513275823 content-margin"
        >
          <MarkdownRender
            source="## The Notable Nteract Notebook

**It's a notebook!**
"
          />
        </div>
        <JSXStyle
          css=".content-margin.jsx-2513275823{padding-left:calc(var(--prompt-width,50px) + 10px);padding-top:10px;padding-bottom:10px;padding-right:10px;}"
          styleId="2513275823"
        />
      </Cell>
      <Cell
        _hovered={false}
        isSelected={false}
        key="15"
      >
        <PapermillView
          status={null}
        />
        <Input
          hidden={false}
        >
          <Prompt
            blank={false}
            counter={11}
            queued={false}
            running={false}
          />
          <Source
            className="input"
            language="ipython"
            theme="light"
          >
            import IPython

from IPython.display import HTML
from IPython.display import Markdown
from IPython.display import display

display(HTML('&lt;h1&gt;Multiple&lt;/h1&gt;'))
display(HTML('&lt;p&gt;Display Elements&lt;/p&gt;'))
display(Markdown('**awesome**'))

print('hey')
42
          </Source>
        </Input>
        <Outputs
          expanded={true}
          hidden={false}
        >
          <Display
            displayOrder={
              Array [
                "application/vdom.v1+json",
                "application/json",
                "application/javascript",
                "text/html",
                "text/markdown",
                "text/latex",
                "image/svg+xml",
                "image/gif",
                "image/png",
                "image/jpeg",
                "text/plain",
              ]
            }
            expanded={false}
            isHidden={false}
            models={Immutable.Map {}}
            outputs={
              Array [
                Object {
                  "data": Object {
                    "text/plain": "<IPython.core.display.HTML object>",
                  },
                  "metadata": Object {},
                  "output_type": "display_data",
                },
              ]
            }
            theme="light"
            transforms={
              Object {
                "application/javascript": [Function],
                "application/json": [Function],
                "application/vdom.v1+json": [Function],
                "image/gif": [Function],
                "image/jpeg": [Function],
                "image/png": [Function],
                "image/svg+xml": [Function],
                "text/html": [Function],
                "text/latex": [Function],
                "text/markdown": [Function],
                "text/plain": [Function],
              }
            }
          />
        </Outputs>
      </Cell>
    </Cells>
    <style>
      :root {
          

  --theme-app-bg: white;
  --theme-app-fg: var(--nt-color-midnight);
  --theme-app-border: var(--nt-color-grey-light);

  --theme-primary-bg: var(--nt-color-grey-lightest);
  --theme-primary-bg-hover: var(--nt-color-grey-lighter);
  --theme-primary-bg-focus: var(--nt-color-grey-light);

  --theme-primary-fg: var(--nt-color-midnight-light);
  --theme-primary-fg-hover: var(--nt-color-midnight);
  --theme-primary-fg-focus: var(--theme-app-fg);

  --theme-secondary-bg: var(--theme-primary-bg);
  --theme-secondary-bg-hover: var(--theme-primary-bg-hover);
  --theme-secondary-bg-focus: var(--theme-primary-bg-focus);

  --theme-secondary-fg: var(--nt-color-midnight-lighter);
  --theme-secondary-fg-hover: var(--nt-color-midnight-light);
  --theme-secondary-fg-focus: var(--theme-primary-fg);

  --theme-primary-shadow-hover: 1px  1px 3px rgba(0, 0, 0, 0.12), -1px -1px 3px rgba(0, 0, 0, 0.12);
  --theme-primary-shadow-focus: 3px  3px 9px rgba(0, 0, 0, 0.12), -3px -3px 9px rgba(0, 0, 0, 0.12);

  --theme-title-bar-bg: var(--theme-primary-bg-hover);

  --theme-menu-bg: var(--theme-primary-bg);
  --theme-menu-bg-hover: var(--theme-primary-bg-hover);
  --theme-menu-bg-focus: var(--theme-primary-bg-focus);
  --theme-menu-shadow: var(--theme-primary-shadow-hover);

  --theme-menu-fg: var(--theme-app-fg);
  --theme-menu-fg-hover: var(--theme-app-fg);
  --theme-menu-fg-focus: var(--theme-app-fg);

  --theme-cell-bg: var(--theme-app-bg);
  --theme-cell-shadow-hover: var(--theme-primary-shadow-hover);
  --theme-cell-shadow-focus: var(--theme-primary-shadow-focus);

  --theme-cell-prompt-bg: var(--theme-primary-bg);
  --theme-cell-prompt-bg-hover: var(--theme-primary-bg-hover);
  --theme-cell-prompt-bg-focus: var(--theme-primary-bg-focus);

  --theme-cell-prompt-fg: var(--theme-secondary-fg);
  --theme-cell-prompt-fg-hover: var(--theme-secondary-fg-hover);
  --theme-cell-prompt-fg-focus: var(--theme-secondary-fg-focus);

  --theme-cell-toolbar-bg: var(--theme-primary-bg);
  --theme-cell-toolbar-bg-hover: var(--theme-primary-bg-hover);
  --theme-cell-toolbar-bg-focus: var(--theme-primary-bg-focus);

  --theme-cell-toolbar-fg: var(--theme-secondary-fg);
  --theme-cell-toolbar-fg-hover: var(--theme-secondary-fg-hover);
  --theme-cell-toolbar-fg-focus: var(--theme-secondary-fg-focus);

  --theme-cell-menu-bg: var(--theme-primary-bg);
  --theme-cell-menu-bg-hover: var(--theme-primary-bg-hover);
  --theme-cell-menu-bg-focus: var(--theme-primary-bg-focus);

  --theme-cell-menu-fg: var(--theme-primary-fg);
  --theme-cell-menu-fg-hover: var(--theme-primary-fg-hover);
  --theme-cell-menu-fg-focus: var(--theme-primary-fg-focus);

  --theme-cell-input-bg: var(--theme-secondary-bg);
  --theme-cell-input-fg: var(--theme-app-fg);

  --theme-cell-output-bg: var(--theme-app-bg);
  --theme-cell-output-fg: var(--theme-primary-fg);

  --theme-cell-creator-bg: var(--theme-app-bg);

  --theme-cell-creator-fg: var(--theme-secondary-fg);
  --theme-cell-creator-fg-hover: var(--theme-secondary-fg-hover);
  --theme-cell-creator-fg-focus: var(--theme-secondary-fg-focus);

  --theme-pager-bg: #fafafa;

  --cm-background: #fafafa;
  --cm-color: black;

  --cm-gutter-bg: white;

  --cm-comment: #a86;
  --cm-keyword: blue;
  --cm-string: #a22;
  --cm-builtin: #077;
  --cm-special: #0aa;
  --cm-variable: black;
  --cm-number: #3a3;
  --cm-meta: #555;
  --cm-link: #3a3;
  --cm-operator: black;
  --cm-def: black;

  --cm-activeline-bg: #e8f2ff;
  --cm-matchingbracket-outline: grey;
  --cm-matchingbracket-color: black;

  --cm-hint-color: var(--cm-color);
  --cm-hint-color-active: var(--cm-color);
  --cm-hint-bg: var(--theme-app-bg);
  --cm-hint-bg-active: #abd1ff;

  --status-bar: #eeedee;

            --theme-cell-shadow-hover: none;
            --theme-cell-shadow-focus: none;
            --theme-cell-prompt-bg-hover: var(--theme-cell-prompt-bg);
            --theme-cell-prompt-bg-focus: var(--theme-cell-prompt-bg);
            --theme-cell-prompt-fg-hover: var(--theme-cell-prompt-fg);
            --theme-cell-prompt-fg-focus: var(--theme-cell-prompt-fg);
          }
        
    </style>
  </div>
</Provider>
`;
